File: Dynamic_Programming/1048. Longest String Chain.cpp
/*
1048. Longest String Chain
*/

// TC:O(nlgn + n*k) SC:O(n)
class Solution {
    static bool cmp(string& a, string& b){
        return a.size() < b.size();
    }
public:
    int longestStrChain(vector<string>& words) {
        sort(words.begin(), words.end(), cmp);
        
        unordered_map<string,int> dp;
        int max_chain = 0;
        for(const auto &word:words){
            dp[word] =1;//從這個字為出發點開始構築
            for(int i = 0; i < word.size(); i++){
                string cur = word.substr(0,i) + word.substr(i+1);//每次都忽略一個字元
                if(dp.find(cur) != dp.end()){
                    dp[word] = max(dp[word], dp[cur]+1);
                } 
            }
            max_chain = max(max_chain, dp[word]);
        }

        return max_chain;
    }
};

File: Dynamic_Programming/1048. Longest String Chain.py
# TC:O(nlgn + n * k * k) SC:O(n * k) 可以不排序
class Solution:
    def longestStrChain(self, words: List[str]) -> int:
        '''
        n^2 * k 大致上才能通過 問說最長長度
        '''
        n = len(words)
        f = defaultdict(int)
        words.sort(key=lambda x: len(x))
        @cache
        def dfs(i:int) -> int:
            m =len(words[i])
            ans = 1
            for j in range(m):
                s = words[i][:j] + words[i][j+1:]
                if s in f:
                    ans = max(ans, 1 + dfs(f[s]))
            f[words[i]] = i
            return ans
        return max(dfs(i) for i in range(n))
        '''
        for word in words:
            f[word] = 1
            m = len(word)
            for i in range(m):
                s = word[:i] + word[i+1:]
                if s in f and f[s] + 1 > f[word]:
                    f[word] = f[s] + 1
            if f[word] > ans:
                ans = f[word]
        return ans
        '''
